### PR TITLE
Always use room state to get display names in Matrix connector

### LINKED ISSUE #1720: Improve performance of room_specific_nicks in matrix connector
# Description

At the moment we iterate over all the members in a room to calculate the room specific nick names:

https://github.com/opsdroid/opsdroid/blob/623a70bbc002db570915fb8336ddd5ed4abb8638/opsdroid/connector/matrix/connector.py#L325-L333

This is very inefficient when we can use the state endpoint to query it for the specific user: https://matrix.org/docs/spec/client_server/r0.5.0#get-matrix-client-r0-rooms-roomid-state-eventtype-statekey

Using the https://matrix-nio.readthedocs.io/en/latest/nio.html#nio.AsyncClient.room_get_state_event method from nio

### PR DESCRIPTION
**Pull request by Cadair**
# Description

This is based on #1715 for the test infrastructure
 
Fixes #1720 

## Status
**ON HOLD**

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by SolarDrew — COMMENTED** [review]
All looks good to me, but there seems to be some overlap with #1715 which I assume will go away once that's merged and pulled, so obviously this wants to wait on that.

File: opsdroid/connector/matrix/connector.py [hunk 1]
```diff
@@ -354,25 +357,17 @@ async def get_nick(self, roomid, mxid):
...
+        )
+        if isinstance(room_state, nio.RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error during getting display name from room state: {room_state.message} (status code {room_state.status_code})"
+            )
             return mxid
-        return res.displayname
+
+        displayname = room_state.content.get("displayname", mxid)
+        return displayname or mxid
```
Comment (Cadair) [inline opsdroid/connector/matrix/connector.py:370]:
```suggestion
        return room_state.content.get("displayname", mxid)
```

File: opsdroid/connector/matrix/tests/test_event_parsing.py [hunk 2]
```diff
@@ -53,4 +56,94 @@ async def test_receive_message(opsdroid, connector_connected, mock_api, caplog):
...
+            "Error during getting display name from room state: unknown error (status code None)",
+        )
+    ]
+
+
+@pytest.mark.add_response(
+    "/_matrix/client/r0/rooms/!12345:localhost/state/m.room.member/@test:localhost",
+    "GET",
+    {"displayname": "test"},
+)
```
Comment (Cadair) [inline opsdroid/connector/matrix/tests/test_event_parsing.py:97]:
```suggestion
```

File: opsdroid/connector/matrix/connector.py [hunk 3]
```diff
@@ -354,25 +357,16 @@ async def get_nick(self, roomid, mxid):
...
+            roomid, "m.room.member", mxid
+        )
+        if isinstance(room_state, nio.RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error during getting display name from room state: {room_state.message} (status code {room_state.status_code})"
+            )
             return mxid
-        return res.displayname
+
+        return room_state.content.get("displayname", mxid)
```
Comment (Cadair) [inline opsdroid/connector/matrix/connector.py:369]:
```suggestion
        return room_state.content.get("displayname", mxid) or mxid
```

File: opsdroid/connector/matrix/connector.py [hunk 4]
```diff
@@ -354,25 +357,16 @@ async def get_nick(self, roomid, mxid):
...
+            roomid, "m.room.member", mxid
+        )
+        if isinstance(room_state, nio.RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error during getting display name from room state: {room_state.message} (status code {room_state.status_code})"
+            )
             return mxid
-        return res.displayname
+
+        return room_state.content.get("displayname", mxid) or mxid
```
[the remark this answers is not in the document]
Comment (Cadair) [reply opsdroid/connector/matrix/connector.py:369]: The or is catch the edge case where the displayname is an empty string. I am not really sure if that is allowed, but I see no reason why it couldn't happen (even if not strictly allowed by the spec) so it seemed like a good hedge.

**Review by SolarDrew — APPROVED** [review]

**Pull request merged**